languages/csharp/Bitwarden.Sdk/BitwardenLibrary.cs

```diff
@@ -2,20 +2,47 @@
 
 namespace Bitwarden.Sdk;
 
-internal static class BitwardenLibrary
+internal static partial class BitwardenLibrary
 {
-    [DllImport("bitwarden_c", CallingConvention = CallingConvention.Cdecl)]
-    private static extern BitwardenSafeHandle init(string settings);
+    [LibraryImport("bitwarden_c", StringMarshalling = StringMarshalling.Utf8)]
+    private static partial BitwardenSafeHandle init(string settings);
 
-    [DllImport("bitwarden_c", CallingConvention = CallingConvention.Cdecl)]
-    private static extern void free_mem(IntPtr handle);
+    [LibraryImport("bitwarden_c", StringMarshalling = StringMarshalling.Utf8)]
+    private static partial void free_mem(IntPtr handle);
 
-    [DllImport("bitwarden_c", CallingConvention = CallingConvention.Cdecl)]
-    private static extern string run_command(string json, BitwardenSafeHandle handle);
+    [LibraryImport("bitwarden_c", StringMarshalling = StringMarshalling.Utf8)]
+    private static partial string run_command(string json, BitwardenSafeHandle handle);
+
+    internal delegate void OnCompleteCallback(IntPtr json);
+
+    [LibraryImport("bitwarden_c", StringMarshalling = StringMarshalling.Utf8)]
+    private static partial void run_command_async(string json, BitwardenSafeHandle handle, OnCompleteCallback on_completed_callback);
 
     internal static BitwardenSafeHandle Init(string settings) => init(settings);
 
     internal static void FreeMemory(IntPtr handle) => free_mem(handle);
 
     internal static string RunCommand(string json, BitwardenSafeHandle handle) => run_command(json, handle);
+
+    internal static Task<string> RunCommandAsync(string json, BitwardenSafeHandle handle, CancellationToken token = default)
+    {
+        var tcs = new TaskCompletionSource<string>(TaskCreationOptions.RunContinuationsAsynchronously);
+
+        try
+        {
+            run_command_async(json, handle, (resultPointer) =>
+            {
+                var stringResult = Marshal.PtrToStringUTF8(resultPointer);
+                tcs.SetResult(stringResult);
+            });
+        }
+        catch (Exception ex)
+        {
+            tcs.SetException(ex);
+        }
+
+        // TODO: Register cancellation on token
```

Comment:
todo?

Reply:
It's still something I want to try out but I think it can be a future enhancement. 

It would require boxing the pointer to the returned `JoinHandle` on the rust side. Sending it back to C# to that we can register a cancellation action and then we likely have to add two new methods on the rust side. Probably `abort_and_drop_handle` and `drop_handle`. Do you have any feelings about that?

Reply:
I ended up doing this, let me know what you think about the implementation.